exonum/src/storage/proof_map_index/key.rs

```diff
@@ -75,52 +69,117 @@ impl ::std::ops::Not for ChildKind {
     }
 }
 
-impl DBKey {
-    /// Create a new bit slice from the given binary data.
-    pub fn leaf<K: ProofMapKey>(key: &K) -> DBKey {
+/// A structure that represents paths to the any kinds of `ProofMapIndex` nodes.
+#[derive(Copy, Clone)]
+pub struct ProofPath {
+    bytes: [u8; PROOF_PATH_SIZE],
+    start: u16,
+}
+
+impl ProofPath {
+    /// Create a path from the given key.
+    pub fn new<K: ProofMapKey>(key: &K) -> ProofPath {
         debug_assert_eq!(key.size(), KEY_SIZE);
 
-        let mut data = [0; KEY_SIZE];
-        key.write(&mut data);
-        DBKey {
-            data: data,
-            from: 0,
-            to: (KEY_SIZE * 8) as u16,
-        }
+        let mut data = [0; PROOF_PATH_SIZE];
+        data[0] = LEAF_KEY_PREFIX;
+        key.write(&mut data[1..KEY_SIZE + 1]);
+        data[PROOF_PATH_LEN_POS] = 0;
+        ProofPath::from_raw(data)
     }
 
-    #[doc(hidden)]
-    pub fn from(&self) -> u16 {
-        self.from
+    /// Shows the type of path.
+    pub fn is_leaf(&self) -> bool {
+        self.bytes[0] == LEAF_KEY_PREFIX
     }
 
-    // TODO: terrible hack, try to remove this (ECR-22)
-    #[doc(hidden)]
-    pub fn set_from(&mut self, from: u16) {
-        self.from = from
+    /// Returns the byte representation of `ProofPath`.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.bytes
     }
 
-    #[doc(hidden)]
-    pub fn to(&self) -> u16 {
-        self.to
+    /// Constructs the `ProofPath` from raw bytes.
+    fn from_raw(raw: [u8; PROOF_PATH_SIZE]) -> ProofPath {
+        debug_assert!(
+            if raw[PROOF_PATH_KIND_POS] == LEAF_KEY_PREFIX {
+                raw[PROOF_PATH_LEN_POS] == 0
+            } else {
+                true
+            },
+            "ProofPath is inconsistent"
+        );
+
+        ProofPath {
+            bytes: raw,
+            start: 0,
+        }
     }
 
-    /// Returns length of the `DBKey`.
-    pub fn len(&self) -> u16 {
-        self.to - self.from
+    /// Sets the right border of the bit range.
+    fn set_end(&mut self, end: u16) {
+        let max_len = (KEY_SIZE * 8) as u16;
```

Comment:
It seems this can be changed to a `const`.

Reply:
I am not sure about this whole message: looks like the valid range for `end` is `0, 1, 2..., 255, 256`, which is much lower than `u16`. Could we use `Option<u8>` for `end`?

Reply:
@alekseysidorov Would it make sense to change `u16` to `Option<u8>` (or possibly, a special type alias, like `ProofPathLength` *everywhere* in the `BitsRange` interface, where `0...256` values should be accepted? Or do you think that would be an overkill?

Reply:
u16 is only used to simplify the code. For example for `start` `u8` is enough, but is not enough for `end` or `len`. 
So I do not want to write code like following:
```rust
    fn len(&self) -> u16 {
        self.end() - self.start() as u16
    }
```
In this way i consider that `Option<u8>` everywhere is overkill.